Launchpad: Migrate Newsletter setup flow

gabrielcaires opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, to avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #86647

## Proposed Changes
Migrate the tasks that are part of the onboarding flow `newsletter`.

**NOTE**:  This flow contains specialized tasks, so I can migrate other smaller flows to reduce the PR size.

<img width="1434" alt="image" src="https://github.com/Automattic/wp-calypso/assets/38718/3e1bf205-5158-4478-b60b-f0d71956ab3d">


## Migrate the tasks

| TASK | MIGRATED? |
|--------|--------|
| 'setup_newsletter' | YES |
|  'plan_selected' |  Previously Migrated |
|  'subscribers_added' | YES |
| 'verify_email' | YES|
| 'set_up_payments' | Previously Migrated |
| 'newsletter_plan_created' | YES |
| 'migrate_content' | YES  |
| 'first_post_published_newsletter' | YES |

### Expected State from feature flags

<img width="1348" alt="Screenshot 2024-01-22 at 17 36 17" src="https://github.com/Automattic/wp-calypso/assets/38718/59cdcfcf-62f2-4ece-847c-97c66b18b642">

Use `yarn feature-search launchpad/new-task-definition-parser/newsletter` to verify

Requires the feature flags:
* `launchpad/new-task-definition-parser`
* `launchpad/new-task-definition-parser/newsletter`

`&flags=launchpad/new-task-definition-parser/newsletter,launchpad/new-task-definition-parser` 

The flags can be enabled appendings the `?flags=` flag on the URL (Please always check if the flags= param is still on the URL when the system redirects back the user from some "no launchpad" page ).

## Testing Instructions

### Free flow
* Start the flow `newsletter`  (setup/newsletter)
* Click 'Launch my newsletter'
* Set a name and description
* Choose Free plan
* Selected a domain
* Select 'Free' again
* Enable the feature flags via URL and reload
* Check if there are console log messages for the following tasks:
<img width="949" alt="image" src="https://github.com/Automattic/wp-calypso/assets/38718/903c6cdc-c338-498f-8bd6-372b11a08e03">

### Paid Flow
* Same as free plan but choosing Paid newsletter
* Selected a domain
* Select a paid plan 
* On the checkout page,  select "Assign a payment method later". It only will be available if we you have WP credits. You can add using our internal tools
* Enable the toggles and check the logs like these.
<img width="819" alt="image" src="https://github.com/Automattic/wp-calypso/assets/38718/30440006-3bb5-4849-a518-e3552767f5a5">\
* Test all launchpad tasks, setting values to completed and not completed tasks

### Migrate Flow 
* Same as free plan but choosing 'Import an existing newsletter'
* Selected a domain
* Select a paid plan 
* On the checkout page,  select "Assign a payment method later". It only will be available if we you have WP credits. You can add using our internal tools
* Enable the toggles and check the logs like these.
* Test all launchpad tasks, setting values to completed and not completed tasks
<img width="695" alt="image" src="https://github.com/Automattic/wp-calypso/assets/38718/836b4736-2b4d-4d65-93ee-dd0d4aaef0f6">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

gabrielcaires committed:
Launchpad: Use new task definition via feature flag

gabrielcaires committed:
Move the deprecated task definition to another file

Fix missing data

gabrielcaires committed:
Move getPlanTaskSubtitle to its own context

gabrielcaires committed:
Move getIsLaunchSiteTaskDisable to its own context

gabrielcaires committed:
Move getLaunchSiteTaskTitle to its own context

gabrielcaires committed:
Move completeLaunchTask to its own context

gabrielcaires committed:
Review tracking on new task definitions

Fix missing tracking

gabrielcaires committed:
Remove unused task ids

gabrielcaires committed:
Unify Task definition

Unify Task definition

gabrielcaires committed:
Fix linter issue

gabrielcaires committed:
Move setup_blog to use the new flow

gabrielcaires committed:
Add design_completed task

gabrielcaires committed:
Move newsletter tasks to use the new structure

gabrielcaires committed:
Move verif_email task

gabrielcaires committed:
Migrate design steps related to the free flow

gabrielcaires committed:
Migrate domain_upsell to use the new structure

gabrielcaires committed:
Migrate first_post_published task

gabrielcaires committed:
Migrate site launch to use new structure

gabrielcaires committed:
Migrate plan_selected to use new structure

gabrielcaires committed:
Migrate plan_complete to use new structure

gabrielcaires committed:
Merge branch 'trunk' into update/refactor-start-writing-flow

gabrielcaires committed:
Move setup_blog to use new structure

gabrielcaires committed:
Move blog_launched to use the new structure

gabrielcaires committed:
Merge branch 'trunk' into update/refactor-start-writing-flow

gabrielcaires committed:
Enable to check the feature flag value by flow

gabrielcaires committed:
Remove extra console.log

gabrielcaires committed:
Merge branch 'update/enable-to-check-flag-by-setup-flow' into update/refactor-start-writing-flow

gabrielcaires committed:
Merge branch 'trunk' into update/enable-to-check-flag-by-setup-flow

gabrielcaires committed:
Merge branch 'trunk' into update/refactor-start-writing-flow

gabrielcaires committed:
Merge branch 'update/enable-to-check-flag-by-setup-flow' into update/refactor-start-writing-flow

gabrielcaires committed:
Enable design-first to use the new task definition

gabrielcaires committed:
Merge branch 'update/refactor-design-first-flow' into update/design-setup-newsletter

gabrielcaires committed:
Remove unused file

gabrielcaires committed:
Add feature flags

gabrielcaires committed:
Migrate tasks

gabrielcaires committed:
Disable start writing

gabrielcaires committed:
Disable start-writing flow in production

gabrielcaires committed:
Merge branch 'update/enable-to-check-flag-by-setup-flow' into update/refactor-start-writing-flow

gabrielcaires committed:
Merge branch 'update/refactor-start-writing-flow' into update/refactor-design-first-flow

gabrielcaires committed:
Rename const name to describe better the purpose

gabrielcaires committed:
Rename const name to describe better the purpose

gabrielcaires committed:
Turn off the star-writing flow flag

gabrielcaires committed:
Turn off the star-writing flow flag

gabrielcaires committed:
Turn on the feature on all development envs

gabrielcaires committed:
Add 'type'  to all types imports

gabrielcaires committed:
Add missing type definitions

gabrielcaires committed:
Merge branch 'update/enable-to-check-flag-by-setup-flow' into update/refactor-start-writing-flow

gabrielcaires committed:
Merge branch 'update/refactor-start-writing-flow' into update/refactor-design-first-flow

gabrielcaires committed:
Enable flow in test

gabrielcaires committed:
Remove newsletter context

gabrielcaires committed:
Merge branch 'update/refactor-design-first-flow' into update/design-setup-newsletter

gabrielcaires committed:
Move subscribers_added to use the new structure

gabrielcaires committed:
Migrate set_up_payments to use the new structure

gabrielcaires committed:
M[127;2uMigrate newsletter_plan_create to use new structure

gabrielcaires committed:
Migrate migrate_content task to use new structure

gabrielcaires committed:
Migrate setup_link_in_bio task to use new structure

gabrielcaires committed:
Migrate link_in_bio_launched task to use the new structure

gabrielcaires committed:
Move links_added task to use the structure

gabrielcaires committed:
Merge branch 'trunk' into update/refactor-design-first-flow

gabrielcaires committed:
Merge branch 'trunk' into update/refactor-design-first-flow

gabrielcaires committed:
Enable feature in staging

gabrielcaires committed:
Turn on feature

gabrielcaires committed:
Merge branch 'trunk' into update/design-setup-newsletter

gabrielcaires committed:
Merge branch 'trunk' into update/design-setup-newsletter

gabrielcaires committed:
Merge branch 'trunk' into update/refactor-design-first-flow

gabrielcaires committed:
Merge branch 'update/refactor-design-first-flow' into update/design-setup-newsletter

gabrielcaires committed:
Add toggle on all envs

gabrielcaires committed:
Add missing set_up_payments

gabrielcaires committed:
Merge branch 'update/design-setup-newsletter' into update/refactor-link-in-bio-flow

gabrielcaires committed:
Add toggle to link-in-bio flow

markbiek commented:
Free plan: Working as expected!
Migrate plan: Working as expected!

Paid plan: I see the correct console messages but clicking the "Set up payment method" task doesn't do anything.


https://github.com/Automattic/wp-calypso/assets/917632/55df42f7-6a83-40f4-b90b-c09da4cb7edd

gabrielcaires commented:
> Paid plan: I see the correct console messages but clicking the "Set up payment method" task doesn't do anything.
@markbiek 
Thanks! | fixed it here and added tests  https://github.com/Automattic/wp-calypso/pull/86852/commits/62749b85dbe1dff9661254e7387305c868694246

I am still investigating how to add tests to the most complex tasks, but since this one is simple I already added it here.

gabrielcaires committed:
Migrante content tasks

gabrielcaires committed:
Migrate email tasks

gabrielcaires committed:
Migrate payment tasks

gabrielcaires committed:
Migrate plan task

gabrielcaires committed:
Migrate post tasks

gabrielcaires committed:
Migrate setup tasks

gabrielcaires committed:
Miigrate subscribers tasks

gabrielcaires committed:
Use the migrated tasks

gabrielcaires committed:
Add feature toggles

gabrielcaires committed:
Add tests to payment task

markbiek reviewed (approved):
LGTM!

gabrielcaires committed:
Merge branch 'update/refactor-link-in-bio-flow' into update/refactor-newsletter-flow
